### ISSUE TITLE
Inconsistencies between TF 1.13.1 and TF 2.4.1 especially on small training splits

### ISSUE THREAD

**Issue by M-Wolff** [keras-team/keras#16444]
**System information**.
- Have I written custom code: Yes
- OS Platform and Distribution: Linux CentOS 7.9 and Windows 10
- TensorFlow installed from: binary (pip install ...)
- TensorFlow version: 2.4.1 vs. 1.13.1
- Python version: 3.7.9
- GPU model and memory: multiple (V100, RTX 2080, Titan XP, ...) on an HPC Cluster at university
- Exact command to reproduce: -

**Describe the problem**.

I analyzed plenty of different parameters for training (size of train-split, training from scratch vs. using pre-trained networks, different networks: ResNet-50 and InceptionV3, ...) on 5 different datasets, which resulted in 3770 neural networks to be trained. I started doing this in TF 2.4.1 and the results were not as expected and in some cases pretty bad (classification performance as bad as randomly guessing), so I trained using the same exact code, datasets and combination of other parameters in TF 1.13.1 and got better results. I already opened an [issue in the tensorflow Github-Repository](https://github.com/tensorflow/tensorflow/issues/50244) about 10 months ago. People there told me to now open the same issue here for faster resolution (that's one way to get rid of an open ticket...) with a link to the previous issue in Tensorflow's Github-Repository to reference all the insights gained there.

To sum up the "bug" or weird behaviour: Especially when using very small train-splits (only 2-3 images per class for training) it seems like TF 2.4.1 (and most likely also in TF 2.8.0, see the issue linked above) perform way worse than TF 1.13.1 using the same exact code and other parameters.

**Describe the current behavior**.
Newer versions of TF (tested on TF 2.4.1 and TF 2.8) perform bad when training on datasets with a very small train-split (2-3 images per class), while an older version (TF 1.13.1) performs just fine.

I already posted this graph in the mentioned [Tensorflow-Issue](https://github.com/tensorflow/tensorflow/issues/50244): It shows the difference of using TF 2.4.1 vs. TF 1.13.1 on the exact same dataset, parameters and code on the 3770 neural networks I trained.
![AllDatasets](https://user-images.githubusercontent.com/56885160/164305737-6bd26d8a-a0d4-4f58-975e-7bab7460ed9b.png)

Negative values mean TF 1.13.1 is better than TF 2.4.1. And yes, I double-checked: There are cases, where the old version of TF reaches  96% accuracy and the new version only 6%. You can clearly see clusters, where the difference in validation accuracy between both versions is especially large. If you want I can filter this plot further (regarding amount of classes, net-type, ...) or you can check it with the [raw data](https://github.com/M-Wolff/Bachelorarbeit_OvO-Klassifikation/blob/main/Ergebnisse/allModelsLog.txt) yourself. Column-names in this CSV file are **GPU-Type**; **Duration**(minutes); **Batch-Size**; **Learning-Rate**; **Dataset** (Dataset name followed by amount of classes); **img-size** in pixels; **classification-scheme** (OvA and OvO); **Net-Type** (R: Resnet50, I: InveptionV3, IP: Inception with changes, can be ignored); **Scratch vs. Finetune** (Pre-Trained networks); **size of train-split** in percent (100 is full, 10 is 10% of original train-split); **Epochs**; **Fold**; **Version of TF or Torch**; **Train_Loss**; **Train_Acc**; **Test_Loss**; **Test_Acc**.

For further details see Readme [here](https://github.com/M-Wolff/Bachelorarbeit_OvO-Klassifikation/tree/main/Ergebnisse) (only in german, sorry).

**Describe the expected behavior**.
Performance in the progression of TF versions should stay the same or even increase, but never decrease like this.
I would expect to get roughly the same performance running the same code on exactly the same datasets in an older (TF 1.13.1) and newer (TF 2.4.1) version.
When using bigger train-splits this expected behaviour is actually present: For example if you filter just for training runs on Cifar, the graph from above reduces to this (only minor differences between both versions).
![Cifar](https://user-images.githubusercontent.com/56885160/164306134-d27328e4-6ddd-48a8-b193-3dd65e9a1915.png)

**Standalone code to reproduce the issue**.
See the code and output produced [here](https://github.com/M-Wolff/TF-Bug).
`train` contains the dataset used as a small example (sub-set of [Swedish Leaf dataset](https://www.cvl.isy.liu.se/en/research/datasets/swedish-leaf/))
`the codefiles` in the root-directory can be ignored
`full_training/training_comparison.py` Launch this (change the path to the dataset used in line 56 and 60: The path there should point to a folder with one sub-folder for each class containing the images) with either "resnet" or "inception" as first console argument. It will train this network for 200 epochs with the currently installed TensorFlow version and log the results in `results.txt`. 
Folder structure is
- train
    - class1
        - img1.png
        - ...
    - ...
    - classN
- test (same as above)

My results are in [`small_trainset_ts10`](https://github.com/M-Wolff/TF-Bug/tree/main/full_training/small_trainset_ts10) and in [`big_trainset_ts100`](https://github.com/M-Wolff/TF-Bug/tree/main/full_training/big_trainset_ts100). The small train-split contains 2-3 images of the Swedish Leaf Dataset per class for training, the big train-split contains about 10x as many images per class for training.
Using the big train-split both versions of TF perform about as good, with the small train-split the newer version (TF 2.4.1) is not better than randomly guessing (20% accuracy for 5 classes) while the old version still manages to reach >45% accuracy. See the results.txt and results.png in both cases.

If you want to use the same exact images I used, please see the directory tree [here](https://github.com/M-Wolff/Bachelorarbeit_OvO-Klassifikation/blob/main/Datens%C3%A4tze/swedishLeaves3folds5.txt). Relevent folders are `swedishLeaves3folds5/exps/exp1/test/*`, `swedishLeaves3folds5/exps/exp1/train_10/*` and `swedishLeaves3folds5/exps/exp1/train_100/*`.

**Source code / logs**.
See [Github-Repo](https://github.com/M-Wolff/TF-Bug) linked and described above.

Please keep in mind that my work on this topic is now about 8-10 months old, just as the [original issue](https://github.com/tensorflow/tensorflow/issues/50244), so I can't promise that I recall everything correctly by now. The whole history has been logged in that issue though.
I am also lacking the time and access to the HPC cluster by now to carry out any bigger experiments regarding this topic. I mentioned and linked the [results of all the training runs I did](https://github.com/M-Wolff/Bachelorarbeit_OvO-Klassifikation/blob/main/Ergebnisse/allModelsLog.txt) (3770 networks in TF 2.4.1 and TF 1.13.1 each plus 2530 runs in PyTorch = 10.070 runs in total) above.

Is there something neither I nor the employees of Tensorflow did not spot for over 10 months now or are the newer versions of Tensorflow really worse when the train-split is very small?

**Comment by rchao** [comment]
The graph features vertical lines and it seems the lines correspond to specific models, and the others are fine. 

1) Can you provide more information about what those models are and what's different about them?
2) Can you provide a colab snippet that can reproduce this for at least one such model?
3) Can you attempt using `predict()` on a data array for each TF version for one of the problematic models? We'd like to understand if there are behavior differences in the forward pass.

Thanks!

**Comment by M-Wolff** [comment]
The vertical lines correspond to datasets with only few training samples, I think.
In the following I focus on one model that is also used in my Google Colab Notebook:

1. `model = keras.applications.inception_v3.InceptionV3(weights=None, include_top=True, classes=5)`
2. https://colab.research.google.com/drive/1vvgI6D7MI8l5SgIRrkJlh03kSlQYpa4l?usp=sharing

The Google Colab snippet features Codeblocks for setting up the environment, including the download of the dataset.

What I found is, that if you use `import keras` the results in TF 1.13.1 are good (~60% accuracy on 5 classes).
If you use `from tensorflow import keras` the results in the **same** version get bad (~20% accuracy on 5 classes).

In TF 1.13.1 both ways of importing keras work (no errors). In TF 2.4.1 the `import keras` leads to the error `module 'tensorflow' has no attribute 'get_default_graph' ` when trying to load the InceptionV3 Model, so you can't produce the good results in TF 2.4.1.

So as of right now it seems, the version of TF is probably not the cause of differences in accuracy, but the **way of importing keras**.

Please confirm, that you can reproduce the behaviour I described here.

Simply run the [code blocks](https://colab.research.google.com/drive/1vvgI6D7MI8l5SgIRrkJlh03kSlQYpa4l?usp=sharing) for
1. Downloading and splitting the dataset
2. Installing tensorflow-gpu==1.13.1 (maybe a restart of the environment is necessary, it will prompt you to do it)
3. Running the training
   3.a once with `import keras` => good results should show up
   3.b once with `from tensorflow import keras` => bad restuls should be produced

The good results will show up within Epoch 40-70. 
You only need to change the one line (the way of importing keras) to switch between good and bad results.

|   |  TF 1.13.1 |  TF 2.4.1 |
|---|---|---|
| `import keras`  |  Predicted Classes<br>[0 0 0 1 1 1 2 2 2 3 3 3 4 4 4]<br>True classes<br>[0 0 0 1 1 1 2 2 2 3 3 3 4 4 4]<br>Accuracy on Train-Split: <br>100.0<br>Loss on Train-Split: 3.872561137541197e-05<br>Accuracy on Test-Split: 66.40000002384186<br>Loss on Test-Split: <br>2.3712333602905273<br>Using TF Version 1.13.1 | Error:<br>`module 'tensorflow' has no attribute 'get_default_graph'`<br>when trying to load the Model |
| `from tensorflow import keras`  | Predicted Classes<br>[3 3 3 3 3 3 3 3 3 3 3 3 3 3 3]<br>True classes<br>[0 0 0 1 1 1 2 2 2 3 3 3 4 4 4]<br>Accuracy on Train-Split: 20.000000298023224<br>Loss on Train-Split: 12.894475936889648<br>Accuracy on Test-Split: 20.000000298023224<br>Loss on Test-Split: <br>12.89447627258301<br>Using TF Version 1.13.1  | Predicted Classes<br>[3 3 3 3 3 3 3 3 3 3 3 3 3 3 3]<br>True classes<br>[0 0 0 1 1 1 2 2 2 3 3 3 4 4 4]<br>Accuracy on Train-Split: <br>20.000000298023224<br>Loss on Train-Split:<br> 5.98341178894043<br>Accuracy on Test-Split: <br>20.000000298023224<br>Loss on Test-Split:<br> 6.112266269683838<br>Using TF Version 2.4.1  |

**Comment by SuryanarayanaY** [comment]
Hello, Thank you for reporting an issue.

We're currently in the process of migrating the new Keras 3 code base from keras-team/keras-core to keras-team/keras.
Consequently, This issue may not be relevant to the Keras 3 code base . After the migration is successfully completed, feel free to reopen this Issue at keras-team/keras if you believe it remains relevant to the Keras 3 code base. If instead this Issue is a bug or security issue in legacy tf.keras, you can instead report a new issue at [keras-team/tf-keras](https://github.com/keras-team/tf-keras/issues), which hosts the TensorFlow-only, legacy version of Keras.

To know more about Keras 3, please read https://keras.io/keras_core/announcement/
